File: Python/Restore IP Addresses.py
class Solution:
    # @param s, a string
    # @return a list of strings
    def restoreIpAddresses(self, s):
        length = len(s)
        if length < 4 or length >12:
            return []
        ret = []
        for i in range(1,4):
            for j in [i+1, i+2, i+3]:
                for k in [j+1, j+2, j+3]:
                    if k >= length:
                        continue
                    s1 = s[:i]
                    s2 = s[i:j]
                    s3 = s[j:k]
                    s4 = s[k:]
                    
                    strlist = [s1,s2,s3,s4]
                    if self.isValid(strlist):
                        ret.append(".".join(strlist))
        return ret
        
    def isValid(self, strlist):
        for s in strlist:
            if s[0] == "0" and s != "0":
                return False
            if int(s) > 255:
                return False
        return True

File: Java/src/Restore IP Addresses.java
public class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> ret = new ArrayList<>();
        int length = s.length();
        for (int i = 1; i<=3 ; i++){
            for (int j= i+1; j<=i+3; j++){
                for (int k = j+1; k<=j+3; k++){
                    if (k>=length) continue;
                    String s1 = s.substring(0,i);
                    String s2 = s.substring(i,j);
                    String s3 = s.substring(j,k);
                    String s4 = s.substring(k,length);
                    if(validIp(s1) && validIp(s2) && validIp(s3) && validIp(s4)){
                        ret.add(s1 + "." + s2 + "." +s3 + "." +s4 );
                    }
                }
            }
        }
        return ret;
        
    }
    
    public boolean validIp(String s){
        if(s.length() > 1 && s.charAt(0) == '0') return false;
        if (s.length() > 3 || Integer.parseInt(s) > 255) return false;
        return true;
        
    }
}